### PR TITLE
Enable HQL Tests

### PR DESCRIPTION
**Pull request by dzou**
This enables the HQL tests and makes a handful of corrections to the Dialect.

Contributes to #52.

### CODE REVIEW

File: google-cloud-spanner-hibernate-testing/src/test/java/org/hibernate/userguide/hql/HQLTest.java [hunk 1]
```diff
@@ -0,0 +1,2293 @@
...
+        "from Person p " +
+        "where p.name like 'Joe'", Person.class)
+      .getResultList();
+      //end::hql-string-literals-example[]
+    });
+  }
+
+  @Test
+  @Ignore
+  // TODO: Test fails due to quote literals in string: 'Joe''s'; figure out if there
```
Comment (meltsufin) [inline google-cloud-spanner-hibernate-testing/src/test/java/org/hibernate/userguide/hql/HQLTest.java]: It's not clear in this file which parts are original, and which ones you changed. Maybe you can split your changes into a separate commit?

Also, this TODO should probably be an issue we're tracking.

Finally, did you have to delete any tests like the ones involving `= ANY (SELECT ...)`?

Comment (dzou) [reply google-cloud-spanner-hibernate-testing/src/test/java/org/hibernate/userguide/hql/HQLTest.java]: I'll make a separate PR with the test annotated with `@Ignore` to visualize changes

File: google-cloud-spanner-hibernate-testing/src/test/java/org/hibernate/userguide/hql/HQLTest.java [hunk 2]
```diff
@@ -1469,24 +1432,6 @@ public void test_hql_collection_expressions_example_4() {
...
-	public void test_hql_collection_expressions_example_5() {
-		doInJPA( this::entityManagerFactory, entityManager -> {
-			Call call = entityManager.createQuery( "select c from Call c", Call.class).getResultList().get( 0 );
-			Phone phone = call.getPhone();
-			//tag::hql-collection-expressions-example[]
-
-			List<Person> persons = entityManager.createQuery(
-				"select p " +
-				"from Person p " +
-				"where :phone = some elements ( p.phones )", Person.class )
```
Comment (ChengyuanZhao) [inline google-cloud-spanner-hibernate-testing/src/test/java/org/hibernate/userguide/hql/HQLTest.java]: this  `= som elements ( ....` becomes ` = ANY (... ` ?

File: google-cloud-spanner-hibernate-testing/src/test/java/org/hibernate/userguide/hql/HQLTest.java [hunk 3]
```diff
@@ -1510,28 +1455,13 @@ public void test_hql_collection_expressions_example_7() {
...
 
-	@Test
-	public void test_hql_collection_expressions_example_8() {
-		doInJPA( this::entityManagerFactory, entityManager -> {
-			//tag::hql-collection-expressions-example[]
-
-			List<Phone> phones = entityManager.createQuery(
-				"select p " +
-				"from Phone p " +
-				"where current_date() > all elements( p.repairTimestamps )", Phone.class )
```
Comment (ChengyuanZhao) [inline google-cloud-spanner-hibernate-testing/src/test/java/org/hibernate/userguide/hql/HQLTest.java]: I assume this becomes the ` = ALL ( .. ` ?

File: google-cloud-spanner-hibernate-testing/src/test/java/org/hibernate/userguide/hql/HQLTest.java [hunk 4]
```diff
@@ -2041,22 +1947,6 @@ public void test_hql_like_predicate_example_2() {
...
-	@Test
-	public void test_hql_like_predicate_escape_example() {
-
-		doInJPA( this::entityManagerFactory, entityManager -> {
-			//tag::hql-like-predicate-escape-example[]
-			// find any person with a name starting with "Dr_"
-			List<Person> persons = entityManager.createQuery(
-				"select p " +
-				"from Person p " +
-				"where p.name like 'Dr|_%' escape '|'", Person.class )
```
Comment (ChengyuanZhao) [inline google-cloud-spanner-hibernate-testing/src/test/java/org/hibernate/userguide/hql/HQLTest.java]: this tests `LIKE` ?  That is in Cloud Spanner though

File: google-cloud-spanner-hibernate-testing/src/test/java/org/hibernate/userguide/hql/HQLTest.java [hunk 5]
```diff
@@ -2073,25 +1963,6 @@ public void test_hql_between_predicate_example_1() {
...
-	@RequiresDialect(MySQL5Dialect.class)
-	public void test_hql_between_predicate_example_2() {
-
-		doInJPA( this::entityManagerFactory, entityManager -> {
-			//tag::hql-between-predicate-example[]
-
-			List<Person> persons = entityManager.createQuery(
-				"select p " +
-				"from Person p " +
-				"where p.createdOn between '1999-01-01' and '2001-01-02'", Person.class )
```
Comment (ChengyuanZhao) [inline google-cloud-spanner-hibernate-testing/src/test/java/org/hibernate/userguide/hql/HQLTest.java]: What does this `between` translate to ?

File: google-cloud-spanner-hibernate-testing/src/test/java/org/hibernate/userguide/hql/HQLTest.java (hunk 1 above)
Comment (dzou) [reply google-cloud-spanner-hibernate-testing/src/test/java/org/hibernate/userguide/hql/HQLTest.java]: 1. Done.
2. Done. Made this issue to track: https://github.com/GoogleCloudPlatform/google-cloud-spanner-hibernate/issues/55
3. Yes, tests using ANY, ALL, and SOME were deleted.

File: google-cloud-spanner-hibernate-testing/src/test/java/org/hibernate/userguide/hql/HQLTest.java [hunk 6]
```diff
@@ -2330,56 +2200,6 @@ public void test_hql_group_by_example_2() {
 		});
 	}
 
-	@Test
-	@RequiresDialect(H2Dialect.class)
-	@RequiresDialect(PostgreSQL81Dialect.class)
-	@RequiresDialect(MySQL5Dialect.class)
-	public void test_hql_group_by_example_3() {
```
Comment (ChengyuanZhao) [inline google-cloud-spanner-hibernate-testing/src/test/java/org/hibernate/userguide/hql/HQLTest.java]: Would this one also work in Spanner? I know this one specifies it only works for some dialects, but it mentions both postgres and mysql, which feel like pretty general SQL to me

File: google-cloud-spanner-hibernate-testing/src/test/java/org/hibernate/userguide/hql/HQLTest.java [hunk 7]
```diff
@@ -1469,24 +1432,6 @@ public void test_hql_collection_expressions_example_4() {
 		});
 	}
 
-	@Test
```
Comment (meltsufin) [inline google-cloud-spanner-hibernate-testing/src/test/java/org/hibernate/userguide/hql/HQLTest.java]: Instead of deleting tests that don't work, I would just put `@Ignore` on them with a short description of why.

Comment (dzou) [reply google-cloud-spanner-hibernate-testing/src/test/java/org/hibernate/userguide/hql/HQLTest.java]: Done.

File: google-cloud-spanner-hibernate-testing/src/test/java/org/hibernate/userguide/hql/HQLTest.java (hunk 2 above)
Comment (dzou) [reply google-cloud-spanner-hibernate-testing/src/test/java/org/hibernate/userguide/hql/HQLTest.java]: No, turns out `SOME` is [its own keyword](https://www.geeksforgeeks.org/sql-some/). Spanner does not yet support "SOME".

File: google-cloud-spanner-hibernate-testing/src/test/java/org/hibernate/userguide/hql/HQLTest.java (hunk 3 above)
Comment (dzou) [reply google-cloud-spanner-hibernate-testing/src/test/java/org/hibernate/userguide/hql/HQLTest.java]: Yup, correct.

File: google-cloud-spanner-hibernate-testing/src/test/java/org/hibernate/userguide/hql/HQLTest.java (hunk 4 above)
Comment (dzou) [reply google-cloud-spanner-hibernate-testing/src/test/java/org/hibernate/userguide/hql/HQLTest.java]: This one doesn't work because it uses `like` with `escape` to specify a custom character to use as an escape delimiter. The escape half is not yet supported in Spanner.

File: google-cloud-spanner-hibernate-testing/src/test/java/org/hibernate/userguide/hql/HQLTest.java (hunk 5 above)
Comment (dzou) [reply google-cloud-spanner-hibernate-testing/src/test/java/org/hibernate/userguide/hql/HQLTest.java]: Seems like it's just the the vanilla Between keyword: https://www.w3schools.com/sql/sql_between.asp

It works in Spanner.

File: google-cloud-spanner-hibernate-testing/src/test/java/org/hibernate/userguide/hql/HQLTest.java (hunk 6 above)
Comment (dzou) [reply google-cloud-spanner-hibernate-testing/src/test/java/org/hibernate/userguide/hql/HQLTest.java]: Hmm it didn't work; seemed to fail at the group by. This one looks like it might be able to make work with some adjustments to Dialect; created issue to track: https://github.com/GoogleCloudPlatform/google-cloud-spanner-hibernate/issues/56

**Comment by meltsufin** [comment]
@dzou Are we adding all of these unsupported features because of which you're ignoring tests to the tracking issue(s)?
Also, it might be useful for the future to reference the issue(s) in the comments, so that when the feature is supported by Spanner, we can easily find the test to re-enable.

**Comment by dzou** [comment]
> @dzou Are we adding all of these unsupported features because of which you're ignoring tests to the tracking issue(s)?
> Also, it might be useful for the future to reference the issue(s) in the comments, so that when the feature is supported by Spanner, we can easily find the test to re-enable.

Yes, the unsupported features have been added to the tracking issue #47.

Hmm, since there is only one issue #47 tracking unsupported features, we would be adding "#47" everywhere there is an Ignore, so I don't think it would be too useful to add; would just be redundant. There are 2 tests which I have identified that could potentially be fixed, and these have separate issues created for them.

**Comment by meltsufin** [comment]
I think it would be easier to track if we had separate issues rather than lumping everything into #47. We can use a tag to group them.

**Comment by dzou** [comment]
> I think it would be easier to track if we had separate issues rather than lumping everything into #47. We can use a tag to group them.

Done. PTAL.

**Comment by meltsufin** [comment]
Perfect! Thanks @dzou!
